Re-implement JObject into c-style struct

yichoi opened the pull request:
Related issue: #439

* Reference counting
  - `iotjs_jval_t`  : Increase jerry refcount. This type of variable must be destroyed by `iotjs_jval_destroy()`
  - `iotjs_jval_t*` : This type of variable should not be destroyed.

* Added `valgrind_tracer` in `IOTJS_VALIDATED_STRUCT`.
  - To find out missing iotjs_jval_destroy() call
  - Allocate small memory on create(), and release it on destroy(), so if someone does not call destroy(), it will report where that memory was allocated at the first place.
  - With only jerry assertion (e.g. Assertion 'ECMA_STRING_IS_REF_EQUALS_TO_ONE (string_p)' failed), it was hard to find out where exactly jerry_value_release() is missing. (Help wanted if there're some better ways!)

chokobole commented on src/iotjs.cpp:
```diff
@@ -95,50 +95,47 @@ static void CleanupModules() {
 
 
 static bool StartIoTjs(Environment* env) {
   // Initialize jerry null and undefined objects.
```
remove this comment asyou remove Jobject::init()

yichoi replied:
In fact, `iotjs_binding_initialize` does almost the same thing.
I've updated this comment to include more values that `iotjs_binding_initialize` initializes.

chokobole commented on src/iotjs.cpp:
```diff
@@ -167,7 +164,6 @@ static bool StartIoTjs(Environment* env) {
   CleanupModules();
 
   // Release jerry null and undefined objects.
```
ditto

chokobole commented on src/iotjs_binding.cpp:
```diff
@@ -19,538 +19,459 @@
 
-  size_t check = jerry_string_to_char_buffer(_obj_val, jerry_buffer, size);
+iotjs_jval_t iotjs_jval_get_property(const iotjs_jval_t* jobj,
+                                            const char* name) {
```
over indented

chokobole commented on src/iotjs_binding.cpp:
```diff
@@ -587,235 +513,253 @@ void iotjs_jargs_destroy(iotjs_jargs_t* jargs) {
 
-iotjs::JObject* iotjs_jargs_get(iotjs_jargs_t* jargs, uint16_t i) {
-  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_jargs_t, jargs);
+const iotjs_jval_t* iotjs_jargs_get(const iotjs_jargs_t* jargs, uint16_t index){
```
space between ) {

chokobole commented on src/iotjs_handlewrap.h:
```diff
@@ -43,8 +43,8 @@ namespace iotjs {
-  HandleWrap(JObject& jobject, /* Object that connect with the uv handle*/
-             uv_handle_t* handle);
+  HandleWrap(const iotjs_jval_t* jobject, uv_handle_t* handle);
+  // jobject: Object that connect with the uv handle*/
```
needless */

chokobole commented on src/module/iotjs_module_buffer.h:
```diff
@@ -23,24 +23,24 @@
+iotjs_jval_t InitBuffer();
 
 
 // Crate buffer object.
```
Create

chokobole commented on src/module/iotjs_module_tcp.cpp:
```diff
@@ -137,15 +138,15 @@ static void AfterConnect(uv_connect_t* req, int status) {
 
   // Take callback function object.
   //  function afterConnect(status)
```
over space

chokobole commented on src/module/iotjs_module_timer.h:
```diff
@@ -21,15 +21,15 @@
-      , _jcallback(NULL) {
+  explicit TimerWrap(Environment* env, const iotjs_jval_t* jtimer)
+      : HandleWrap(jtimer, reinterpret_cast<uv_handle_t*>(&_handle)) {
     // Initialze timer handler.
```
Initialize

yichoi commented:
@chokobole, Thank you for the review. I've updated the PR as you commented.

And I've added one more modification: rename iotjs_binding.cpp into iotjs_binding.c.

LaszloLango commented on src/iotjs_binding.h:
```diff
@@ -26,309 +26,286 @@
- private:
-  JObject _value;
-  JResultType _type;
+struct iotjs_jval_t;
```
A cannot find where is `iotjs_jval_t` defined

yichoi replied:
This is useless line. iotjs_jval_t is defined in line 57-59, like below
```
typedef struct {
  jerry_value_t value;
} IOTJS_VALIDATED_STRUCT(iotjs_jval_t);
```

zherczeg commented on src/iotjs.cpp:
```diff
@@ -166,8 +163,7 @@ static bool StartIoTjs(Environment* env) {
 
-  // Release jerry null and undefined objects.
-  JObject::cleanup();
+  // Release commonly used jerry values
```
I noticed that the original comments terminated by dot (.), but the new comments does not. Is this an intentional change?

yichoi replied:
No it was not intentional change, I've fixed it while rebasing the commit.

zherczeg commented on src/module/iotjs_module_fs.cpp:
```diff
@@ -63,24 +67,26 @@ static void After(uv_fs_t* req) {
-      default:
+      default: {
         iotjs_jargs_append_null(&jarg);
+      }
```
I would add a break here.

zherczeg commented on src/module/iotjs_module_fs.cpp:
```diff
@@ -196,42 +197,37 @@ JHANDLER_FUNCTION(Read) {
   if (offset >= data_length) {
-    JHANDLER_THROW_RETURN(RangeError, "offset out of bound");
+    JHANDLER_THROW(RANGE, "offset out of bound");
   }
```
Is it a good idea if the second throw overwrites the first throw? Isn't it create a memory leak?

zherczeg commented on src/module/iotjs_module_httpparser.cpp:
```diff
@@ -308,64 +305,64 @@ void HTTPParserWrap::Initialize(http_parser_type type) {
+  uintptr_t handle = iotjs_jval_get_object_native_handle(jparser);
+  return reinterpret_cast<HTTPParserWrap*>(handle);
+}
+
```
I think double newline should be used here

yichoi replied:
I added one more newline here.

zherczeg commented:
Great patch!

yichoi commented:
Updated the commit after reviews.

zherczeg commented:
LGTM
